Location reading interval

matteoKuama opened the pull request:
Stream location now manages also the notifications that keeps to send the locations even if the app is killed. 
It is possibile to set the notification interval and it must be >=60 seconds. 
It is possible to set the accuracy and to customise the notification message content. 
The **UserNotificationDelegate.swift** and **LocationUtility.swift** files are no longer needed.

matteoKuama committed:
Added a new method to wake the app when is backgrounded or killed. The createMonitoredRegions method creates three CLCircularRegions in front, at the left and at the right of the current position and direction of the user. The user will be notified when enters or exits from one of those three regions. The Haversine inverted formula is used to calculate the latitude and the longitude of the position in front of the user. Removed the sound from the notifications.

matteoKuama committed:
When monitoring regions, added also a monitored region around the current position.

matteoKuama committed:
When a new notification appears, it also cancels the previous notifications.

matteoKuama committed:
Added the method didFailWithErrors from the location manager protocol

matteoKuama committed:
Extracted static methods for retrieving a new location from a given point to a new class

matteoKuama committed:
Stream location now manages also the notifications that keeps to send the locations even if the app is killed. It is possibile to set the notification interval and it must be >=60 seconds. It is possible to set the accuracy and to customise the notification message content. The UserNotificationDelegate.swift and LocationUtility.swift files are no longer needed.

kuamanet commented:
@matteoKuama 
This is an example on how you should write the comments:
```
/**
 This method sum two double numbers and returns.

 Here is the discussion. This methods adds two double and return the optional Double.


 - parameter number1: First Double Number.
 - parameter number2: Second Double Number.
 - returns: The sum of two double numbers.

 # Notes: #
 1. Parameters must be **double** type
 2. Handle return type because it is optional.

 # Example #
\`\`\`
 if let sum = self.add(number1: 23, number2: 34) {
  print(sum)
 }
 \`\`\`
*/
```

Also please create constants with reasonable names for the magic numbers  inside `LocationUtility` and evaluate the possibility to transform those static methods into extensions

matteoKuama committed:
Added comments for StreamLocation.swift.

matteoKuama commented:
Added comments for StreamLocation.swift.
UserNotificationDelegate.swift and LocationUtility.swift files are no longer needed, so they can be deleted.

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+ When the app gets killed or suspended it reads the location every 60 seconds minimum.
+ 
+ # Example: #
+ let streamLocation = StreamLocation()
```
```suggestion
\`\`\`
 let streamLocation = StreamLocation()
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+ DispatchQueue.main.async{
+     self.cancellable = stream.subject.publisher.sink{ location in
+      ...   }
  */
```
```suggestion
\`\`\`
 */
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
      */
-    public func startUpdatingLocations(){
-        locationManager.startUpdatingLocation()
+    public func setNotificationContent(title: String?, body: String?){
```
```suggestion
    public func setNotificationContent(title: String? = nil, body: String? = nil){
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
      */
-    public func stopUpdatingLocations(){
-        locationManager.stopUpdatingLocation()
+    public func setAccuracy(accuracy: Accuracy){
```
```suggestion
    public func setAccuracy(_ accuracy: Accuracy){
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
-            for region in monitoredRegions {
-                locationManager.stopMonitoring(for: region)
-            }
+    public func setNotificationTimeInterval(timeInterval: Double){
```
```suggestion
    public func setNotificationTimeInterval(_ timeInterval: Double){
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+     # Example: #
+     \`\`\`
+     let streamLocation = StreamLocation()
+     streamLocation.setNotificationTimeInterval(timeInterval: 90.0)
```
```suggestion
     streamLocation.setNotificationTimeInterval(90.0)
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+    }
+
+    /**
+     This method sets the trigger that will launch a notification.
```
```suggestion
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+    3.timeInterval must be **Double** type.
+    4.repeats must be **Bool** type.
+     */
+    func showNotification(notificationTitle: String, notificationBody: String, timeInterval: Double, repeats: Bool) {
```
```suggestion
    func showNotification(withTitle notificationTitle: String, andBody notificationBody: String, andInterval timeInterval: Double, repeats: Bool) {
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+    
+    /**
+    This method removes both pending and delivered notifications.
+    
```
```suggestion
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+    /**
+     This method starts to register the location of the user.
+     
+     When this method is invoked, it starts to read the location of the user and register a notification that will start to appear when the app goes in background.
```
```suggestion
     When this method is invoked, it starts to read the location of the user and register a notification that will start to appear when the app goes in background.
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+     
+     When this method is invoked, it starts to read the location of the user and register a notification that will start to appear when the app goes in background.
+     
+     #Notes: #
```
```suggestion
     # Notes: #
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+     #Notes: #
+     1. If you want to customize the notification appearance (title, content, timeInterval), you should do it before calling this method.
+     
+     #Example: #
```
```suggestion
     # Example: #
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+    /**
+     This method stops to register the location of the user.
+     
+     When this method is invoked, it stops to read the location of the user when the app is foreground or in the background. To stop the updates also when the app is terminated, you should call stopUpdates.
```
```suggestion
     When this method is invoked, it stops to read the location of the user when the app is foreground or in the background. To stop the updates also when the app is terminated, you should call stopUpdates.
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+     
+     When this method is invoked, it stops to read the location of the user when the app is foreground or in the background. To stop the updates also when the app is terminated, you should call stopUpdates.
+ 
+     #Notes: #
```
```suggestion
     # Notes: #
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+     #Notes: #
+     This method stops to read the location only when the app is in foreground or in the background, **not** when the app is terminated.
+     
+     #Example: #
```
```suggestion
     # Example: #
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+    /**
+     This method stops to register the location of the user and to update the notifications.
+     
+     This method stops to register the location of the user when it is in foreground, background and terminated. It also removes all the notifications, both pending and delivered.
```
```suggestion
This method stops to register the location of the user when it is in foreground, background and terminated. It also removes all the notifications, both pending and delivered.
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+     
+     This method stops to register the location of the user when it is in foreground, background and terminated. It also removes all the notifications, both pending and delivered.
+     
+     #Notes: #
```
```suggestion
     # Notes: #
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+     1.This method completely stops the updates both of notification and location updates.
+     2.This method **must** be invoked when the app should stop, otherwise it will continue to update locations and notifications.
+     
+     #Example: #
```
```suggestion
     # Example: #
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
     /**
-     This method creates a region centered in the current location and it asks to be notified when the user exits from the region. The radius of the region is set to 1mt but the minimum radius is approx. 150mt and it's decided by iOS.
+    This method is called when a new location is available.
+     
```
```suggestion
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+     
+     This method is inherited from the CLLocationManagerDelegate protocol and it is automatically called when a new location is available. If there is a new location, this method publishes it in the stream.
+     
+     #Notes: #
```
```suggestion
     # Notes: #
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+    
+    /**
+     This method is called when an error is thrown during the reading of the location.
+     
```
```suggestion
```

kuamanet commented on ios-location-listener/Classes/UserNotificationDelegate.swift:
```diff
@@ -26,7 +26,7 @@ extension CLLocationManager: UNUserNotificationCenterDelegate {
     
     public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
-        completionHandler([.badge, .sound])
+        completionHandler([.badge)
```
```suggestion
        completionHandler([.badge])
```

kuamanet commented on ios-location-listener/Classes/LocationUtility.swift:
```diff
@@ -0,0 +1,45 @@
+    }
+    
+    /// convert degrees to radians
+    static func deg2rad(_ number: Double) -> Double {
+        return number * .pi / 180
+    }
+
+    /// convert radians to degrees
+    static func rad2deg(_ number: Double) -> Double {
+        return number * 180 / .pi
+    }
```
These shuold be inside a `Double` [extension](https://docs.swift.org/swift-book/LanguageGuide/Extensions.html)

kuamanet commented on ios-location-listener/Classes/LocationUtility.swift:
```diff
@@ -0,0 +1,45 @@
+class LocationUtility {
+    
+    /// Calculate a new point centered 150 meters from location and with angle bearing from location using Haversine inverted formula
+    static func getCheckPointsLocation(location:CLLocation, bearing: Double)->CLLocation{
+        // data
+        let earthRadius = 6371e3 // in meters
+        let distance = 150.0 // in meters
+        let angularDistance = distance/earthRadius
+        let startingLatitude = deg2rad(location.coordinate.latitude)
+        let startingLongitude = deg2rad(location.coordinate.longitude)
+        let mBearings = deg2rad(bearing)
+        
+        // new latitude
+        var newLat = asin(sin(startingLatitude)*cos(angularDistance) +
+                            cos(startingLatitude)*sin(angularDistance)*cos(mBearings))
+        newLat = Double(round(rad2deg(newLat)*10e8)/10e8)
+        // new longitude
+        var newLon = startingLongitude + atan2(sin(mBearings)*sin(angularDistance)*cos(startingLatitude),
+                                               cos(angularDistance)-sin(startingLatitude)*sin(newLat))
+        newLon = Double(round(rad2deg(newLon)*10e8)/10e8)
+        return CLLocation(latitude: newLat, longitude: newLon)
+    }
```
This should be a `CLLocation` [extension](https://docs.swift.org/swift-book/LanguageGuide/Extensions.html) and the `distance` should be a parameter

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
 //
-// StreamLocation.swift
-// StreamLocation
+//  LocationRequest.swift
```
Is it intentional? The file is still named `StreamLocation.swift`

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
 /**
- This class manages to create a stream of location that will continue to update the user even if the app is terminated
+ This class registers the location of the user, even when the app is terminated.
+ 
```
```suggestion
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+    private var notificationBody: String = ""
+    private var notificationTitle: String = "Location Update"
+    private let requestIdentifier = UUID.init().uuidString
+    private var notificationTimeInterval = 60.0 // default value in seconds
```
Create a constant instead of using a magic number

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
+    private let requestIdentifier = UUID.init().uuidString
+    private var notificationTimeInterval = 60.0 // default value in seconds
     private let locationManager = CLLocationManager()
+    private let minimumTimeInterval = 60.0 // in seconds
```
Create a constant instead of using a magic number

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
-     Setup the location manager: max accuracy, asks for backgroundupdates and the permission to retrieve the location when in the state of in use and always.
+     This method setups the location manager.
+
+     This method sets the delegate for the location manager, the accuracy, the distance filter, 
```
```suggestion
     This method sets the delegate for the location manager, the accuracy, the distance filter, 
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
-        if(!checkAuthorization()){
-            print("No location services active, please activate the GPS")
+        if(!checkLocationAuthorization()){
+            print("Provide authorizations")
```
Shouldn't we throw an error here?

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
-     Check if the user has the GPS active
+    This method checks if there are the authorizations to read the location and returns.
+ 
+    This method checks if the locations services are enabled and if it's possible to monitor the location.
```
```suggestion
This method checks if the locations services are enabled and if it's possible to monitor the location.
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
-     It stops the updates of locations
+     This method sets the accuracy for the location updates.
+     
+     This method takes a value for accuracy and set the correspondant value for the location updates.
```
```suggestion
This method takes a value for accuracy and set the correspondant value for the location updates.
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
-     It stops the updates of locations and the monitoring of regions. The monitoring keeps alive the app even if it is closed
+     This method sets the time interval between each notification.
+ 
+     This method sets the time interval between each notification
```
```suggestion
This method sets the time interval between each notification
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
-     Inherited from the CLLocationManagerDelegate. This method is called when there is an update of location.
+    This method requests the authorizations to schedule the launch of a new notifications.
+     
+    This method asks for the authorizations to display notifications and it schedule the launch of a new notification.
```
```suggestion
This method asks for the authorizations to display notifications and it schedule the launch of a new notification.
```

kuamanet commented on ios-location-listener/Classes/StreamLocation.swift:
```diff
@@ -1,130 +1,357 @@
-     Inherited from the CLLocationManagerDelegate.  This method is called when the user exits from a region. If the app was terminated, the call to scheduleLocalNotification reactivates the app to restart the location updates.
+     This method is called before a notification appears.
+     
+     This method ovverrides from the NotificationCenterDelegate protocol. This method is automatically called before the notification appears on the screen. When this method is invoked, it requests for a new location update and it sends it to the stream.
```
```suggestion
This method overrides from the NotificationCenterDelegate protocol. This method is automatically called before the notification appears on the screen. When this method is invoked, it requests for a new location update and it sends it to the stream.
```

matteoKuama committed:
Fixed comments as suggest.
It throws an error when no GPS authorizations are granted.
Created some constants instead of magic numbers.
